### PR TITLE
feat: Support BatchWrite API

### PR DESCRIPTION
**Pull request by SandeepTuniki**
Implementation of the BatchWrite API. See internal link go/cs-batch-write-client-spec for details. This pull request includes the implementation and tests but not the samples (which will appear in https://github.com/GoogleCloudPlatform/ruby-docs-samples).

### CODE REVIEW

File: google-cloud-spanner/lib/google/cloud/spanner/batch_write_results.rb [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+        def error?
+          !ok?
+        end
+
+        ##
+        # A list of the indexes of successful mutation groups.
+        #
+        # @return [::Array<::Integer>]
+        #
+        def ok_indexes
```
Comment (aandreassa) [inline google-cloud-spanner/lib/google/cloud/spanner/batch_write_results.rb:136]: `successful_mutation_indexes` (or indices)? Unless this name was already predetermined.

**Review by aandreassa — APPROVED** [review]

File: google-cloud-spanner/lib/google/cloud/spanner/session.rb [hunk 2]
```diff
@@ -656,6 +657,97 @@ def commit transaction_id: nil, commit_options: nil,
...
+        #
+        def batch_write request_options: nil, call_options: nil
+          ensure_service!
+          b = BatchWrite.new
+          yield b
+          response = service.batch_write path, b.mutation_groups_grpc,
+                                         request_options: request_options,
+                                         call_options: call_options
+          results = BatchWriteResults.new response
+          @last_updated_at = Time.now
```
[the remark this answers is not in the document]
Comment (dazuma) [reply google-cloud-spanner/lib/google/cloud/spanner/session.rb:747]: You're right, this should be changed to use `CLOCK_MONOTONIC`. And yes, let's keep that separate. I filed https://github.com/googleapis/ruby-spanner/issues/109 to track it.

File: google-cloud-spanner/lib/google/cloud/spanner/batch_write_results.rb (hunk 1 above)
Comment (dazuma) [reply google-cloud-spanner/lib/google/cloud/spanner/batch_write_results.rb]: You're right, that's an awkward name. Changed to `successful_indexes`.

File: google-cloud-spanner/lib/google/cloud/spanner/batch_write.rb [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+require "google/cloud/spanner/mutation_group"
+require "google/cloud/spanner/v1"
+
+module Google
+  module Cloud
+    module Spanner
+      ##
+      # Part of the BatchWrite DSL. This object is passed as a parameter to the block
+      # passed to {Google::Cloud::Spanner::Client#batch_write}. Use this object to add
+      # mutation groups to the batch.
```
[the remark this answers is not in the document]
Comment (dazuma) [reply google-cloud-spanner/lib/google/cloud/spanner/batch_write.rb:26]: Done.

**Pull request merged**